### ISSUE TITLE
Clarification on usage of objects with private fields/methods

### ISSUE THREAD

**Issue by MaryamZi** [ballerina-platform/ballerina-spec#306]
**Description:**
Say I have the following object type defined in module `x`.

```ballerina
public type Foo object {
    private int i = 1;

    private function abc() {
    }

    public function def {
    }
}
```

According to the object subtyping rules, it is not possible to define a subtype of this object, since it has private fields/methods.

The question is whether the following scenarios are allowed, given the restriction when there are private fields/methods.

**Example 1:**
```ballerina
public function main() {
    x:Foo f = new;
    x:Foo f2 = f;     // assigning a `x:Foo` value to a variable of the same type
}
```

**Example 2:**
Create an `x:Foo` value in one module and use it as an argument to invoke a function in another module.

Module `y`
```ballerina
public function takeFoo(x:Foo f) {
    // Function accepting an `x:Foo` value.
}
```

Module `z`
```ballerina
public function main() {
    // Creates an `x:Foo` and invokes the `takeFoo()` function in `y`.
    x:Foo f = new;
    y:takeFoo(f);
}
```
 

Basically, if these are valid, a check would have to be done which validates that they are the exact same type (taking into consideration org name, module name, type name, version, etc.), since object subtyping checks will fail for these scenarios.

**Comment by jclark** [comment]
Both your examples are correct. You need to compare the visibility regions of the fields/methods. See the section called "Visibility" in the spec. Feel free to reopen if this is not clear.
